File: House Robber 2/SpaceOptimization.py
def HouseRobber(size,arr):
    prev2 = 0
    prev1 = arr[0]

    for i in range(1,size+1):
        take =  arr[i]
        if i > 1: take += prev2 
        nonTake = prev1

        prev2 = prev1 
        prev1 = max(take,nonTake)
    
    return prev1

if __name__ == '__main__':
    arr = [1,2,3,1]
    size =  len(arr)

    arr1 = arr[:size-1]
    arr2 = arr[1:]

    print( max(HouseRobber(size-2,arr1) , HouseRobber(size-2,arr2)) )

File: House Robber 2/SpaceOptimization.java

class SpaceOptimization {

    public static int HouseRobber(int size,int[] arr) {
        int prev1 = arr[0];
        int prev2 = 0;

        for(int i = 1; i < size; i ++) {
            int take = arr[i];
            if(i > 1) take += prev1; 
            int nonTake = prev2;

            prev2 = prev1;
            prev1 = Math.max(nonTake,take);
        }

        return prev1;
    }

    public static void main(String[] args) {
        int[] nums = {1,2,3,1};
        int size = nums.length;   

        int arr1[] = new int[size-1];
        int arr2[] = new int[size-1];
        int dp1[] = new int[size-1];
        int dp2[] = new int[size-1];

         int k = 0;
        for(int i = 0 ; i < size; i ++ ) {
            if( i > 0) {
                arr2[k] = nums[i];
                k += 1;
            }

            if( i < size-1) arr1[i] = nums[i];
        }


        int num1 = HouseRobber(size-1 ,arr1);
        int num2 = HouseRobber(size-1, arr2);

        System.out.println(Math.max(num1,num2));        
    }
}